packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp

```diff
@@ -8813,21 +8867,204 @@ namespace Tpetra {
+                                                  colind,
+                                                  rowTransfer,
+                                                  MyImporter,
+                                                  MyDomainMap,
+                                                  type3PIDs,
+                                                  type3LIDs,
+                                                  ReducedComm);
+        }
+        Teuchos::ArrayView<const int>  EPID1 =  MyImporter->getExportPIDs();// SourceMatrix->graph->importer
+        Teuchos::ArrayView<const LO>   ELID1 =  MyImporter->getExportLIDs();
+
+        Teuchos::ArrayView<const int>  TEPID2  =  rowTransfer.getExportPIDs(); // row matrix
+        Teuchos::ArrayView<const LO>   TELID2  =  rowTransfer.getExportLIDs();
+
+        const int numCols = getGraph()->getColMap()->getNodeNumElements(); // may be dup
+        // from EpetraExt_MMHelpers.cpp: build_type2_exports
+        std::vector<bool> IsOwned(numCols,true);
+        std::vector<int>  SentTo(numCols,-1);
+        if(!MyImporter.is_null())
+            for(auto && rlid : MyImporter->getRemoteLIDs() ) // the remoteLIDs must be from sourcematrix
+                IsOwned[rlid]=false;
+
+        std::vector<std::pair<int,GO> > usrtg;
+        usrtg.reserve(TEPID2.size());
+
+        for(uint i=0;i<TEPID2.size() ;++i) {
+            const LO  row = TELID2[i];
+            const int pid = TEPID2[i];
+            for(uint j=rowptr[row]; j<rowptr[row+1]; j++) {
+                    int col=colind[j];
+                    if(IsOwned[col] && SentTo[col]!=pid){
+                        SentTo[col]    = pid;
+                        GO gid = this->getColMap()->getGlobalElement(col); // *this is sourcematrix
+                        usrtg.push_back(std::pair<int,GO>(pid,gid));
+                    }
+            }
+        }
+// This sort can _not_ be omitted.[
+        std::sort(usrtg.begin(),usrtg.end()); // default comparator does the right thing, now sorted in gid order
+        auto eopg = std ::unique(usrtg.begin(),usrtg.end());
+        // 25 Jul 2018: Could just ignore the entries at and after eopg.
+        usrtg.erase(eopg,usrtg.end());
+
+        const uint type2_us_size= usrtg.size();
+        Teuchos::ArrayRCP<int>  EPID2=Teuchos::arcp(new int[type2_us_size],0,type2_us_size,true);
+        Teuchos::ArrayRCP< LO>  ELID2=Teuchos::arcp(new  LO[type2_us_size],0,type2_us_size,true);
+
+        int pos=0;
+        for(auto && p : usrtg) {
+            EPID2[pos]= p.first;
+            ELID2[pos]= this->getDomainMap()->getLocalElement(p.second);
+            pos++;
+        }
+
+        Teuchos::ArrayView<int>  EPID3  = type3PIDs();
+        Teuchos::ArrayView< LO>  ELID3  = type3LIDs();
+        GO InfGID = std::numeric_limits<GO>::max();
+        int InfPID = INT_MAX;
+#ifdef TPETRA_MIN3
+#  undef TPETRA_MIN3
+#endif // TPETRA_MIN3
+#define TPETRA_MIN3(x,y,z) ((x)<(y)?(std::min(x,z)):(std::min(y,z)))
```

Comment:
Can you #undef this when you're done with it

Reply:
Done in my latest commit.